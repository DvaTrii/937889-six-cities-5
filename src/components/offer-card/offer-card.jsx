import React from "react";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";
import {setHoveredOfferId, resetHoveredOfferId} from "../../store/action";
import {connect} from "react-redux";

const OfferCard = ({card, cardClass, setHoveredOfferIdAction, resetHoveredOfferIdAction}) => {

  const {id, previewImage, isPremium, isBookmark, price, title, type, rating} = card;

  return (
    <article className={`${cardClass} place-card`}
      onMouseEnter={() => {
        setHoveredOfferIdAction(id);
      }}
      onMouseLeave={() => {
        resetHoveredOfferIdAction();
      }}>
      {isPremium && (
        <div className="place-card__mark">
          <span>Premium</span>
        </div>)}
      <div className="cities__image-wrapper place-card__image-wrapper">
        <a href="#">
          <img className="place-card__image"
            src={previewImage}
            width="260" height="200"
            alt="Place image" />
        </a>
      </div>
      <div className="place-card__info">
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">&euro;{price}</b>
            <span className="place-card__price-text">&#47;&nbsp;night</span>
          </div>
          <button className={`place-card__bookmark-button
           ${isBookmark && `place-card__bookmark-button--active`}
           button`} type="button">
            <svg className="place-card__bookmark-icon" width="18" height="19">
              <use xlinkHref="#icon-bookmark" />
            </svg>
            <span className="visually-hidden">{isBookmark ? `In bookmarks` : `To bookmarks`}</span>
          </button>
        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: `${rating * 20}%`}}/>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <h2 className="place-card__name">
          <Link to={`/offer/${id}`}>{title}</Link>
        </h2>
        <p className="place-card__type">{type}</p>
      </div>
    </article>
  );
};

OfferCard.propTypes = {
  card: PropTypes.shape({
    id: PropTypes.number.isRequired,
    previewImage: PropTypes.string.isRequired,
    isPremium: PropTypes.bool.isRequired,
    isBookmark: PropTypes.bool,
    price: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
  }),
  cardClass: PropTypes.string.isRequired,
  setHoveredOfferIdAction: PropTypes.func.isRequired,
  resetHoveredOfferIdAction: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  hoveredOfferId: state.hoveredOfferId,
});

const mapDispatchToProps = (dispatch) => ({
  setHoveredOfferIdAction(id) {
    dispatch(setHoveredOfferId(id));
  },
  resetHoveredOfferIdAction() {
    dispatch(resetHoveredOfferId());
  }
});

export {OfferCard};
export default connect(mapStateToProps, mapDispatchToProps)(OfferCard);
