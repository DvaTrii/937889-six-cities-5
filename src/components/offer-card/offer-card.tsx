import * as React from "react";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";
import {setHoveredOfferId, resetHoveredOfferId} from "../../store/app/actions";
import {getHoveredOfferId} from "../../store/app/selectors";
import {connect} from "react-redux";
import FavoritesButton from "../favorites-button/favorites-button";
import {CardClass, FavoritesButtonClassPrefix} from "../../const";

const OfferCard = ({card, cardClass, setHoveredOfferIdAction, resetHoveredOfferIdAction}) => {

  const {id, previewImage, isPremium, isBookmark, price, title, type, rating} = card;

  return (
    <article className={`${cardClass}card place-card`}
      onMouseEnter={() => {
        setHoveredOfferIdAction(id);
      }}
      onMouseLeave={() => {
        resetHoveredOfferIdAction();
      }}>
      {isPremium && (
        <div className="place-card__mark">
          <span>Premium</span>
        </div>)}
      <div className={`${cardClass === CardClass.MAIN ? `cities__` : cardClass}image-wrapper place-card__image-wrapper`}>
        <a href="#">
          <img className="place-card__image"
            src={previewImage}
            width={`${cardClass === CardClass.FAVORITE ? 150 : 260}`}
            height={`${cardClass === CardClass.FAVORITE ? 110 : 200}`}
            alt="Place image" />
        </a>
      </div>
      <div className={`${cardClass === CardClass.FAVORITE ? `favorites__card-info` : ``} place-card__info`}>
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">&euro;{price}</b>
            <span className="place-card__price-text">&#47;&nbsp;night</span>
          </div>
          <FavoritesButton
            classPrefix={FavoritesButtonClassPrefix.MAIN}
            isBookmark={isBookmark}
            offerId={id}
          />
        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: `${rating * 20}%`}}/>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <h2 className="place-card__name">
          <Link to={`/offer/${id}`}>{title}</Link>
        </h2>
        <p className="place-card__type">{type}</p>
      </div>
    </article>
  );
};

OfferCard.propTypes = {
  card: PropTypes.shape({
    id: PropTypes.number.isRequired,
    previewImage: PropTypes.string.isRequired,
    isPremium: PropTypes.bool.isRequired,
    isBookmark: PropTypes.bool,
    price: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
  }),
  cardClass: PropTypes.string.isRequired,
  setHoveredOfferIdAction: PropTypes.func.isRequired,
  resetHoveredOfferIdAction: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  hoveredOfferId: getHoveredOfferId(state),
});

const mapDispatchToProps = (dispatch) => ({
  setHoveredOfferIdAction(id) {
    dispatch(setHoveredOfferId(id));
  },
  resetHoveredOfferIdAction() {
    dispatch(resetHoveredOfferId());
  }
});

export {OfferCard};
export default connect(mapStateToProps, mapDispatchToProps)(OfferCard);
