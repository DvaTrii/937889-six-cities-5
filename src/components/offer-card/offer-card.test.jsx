import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter} from "react-router-dom";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";

import {OfferCard} from "./offer-card.tsx";

const mockStore = configureStore([]);

it(`Should OfferCard render properly`, () => {
  const card = {
    id: 101,
    city: {
      name: `Amsterdam`,
      coordinates: { // был [33, 33]
        latitude: 52.38333,
        longitude: 4.9
      }
    },
    coordinates: {
      latitude: 52.3909553943508,
      longitude: 4.85309666406198
    },
    previewImage: `img/apartment-02.jpg`,
    pictures: [`img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`],
    isPremium: true,
    isBookmark: true,
    price: 100,
    title: `Wood and stone place`,
    type: `Apartment`,
    rating: 5,
    description: `Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium`,
    bedroomsMax: `4 Bedrooms`,
    guestsMax: `Max 5 adults`,
    amenities: [`Wi-Fi`, `Heating`, `Kitchen`, `Parking`, `Flowers`, `Dishwasher`, `Towels`, `TV`],
    hostInfo: {
      avatar: `img/avatar-max.jpg`,
      name: `Ross`,
      isSuper: true,
    },
  };

  const cardClass = `cities__place-`;

  const noop = () => {};
  const store = mockStore({});

  const tree = renderer
    .create(
        <Provider store={store}>
          <BrowserRouter>
            <OfferCard
              card={card}
              setHoveredOfferIdAction={noop}
              resetHoveredOfferIdAction={noop}
              cardClass={cardClass}
              getReviewsAction={noop}/>
          </BrowserRouter>
        </Provider>
    )
        .toJSON();
  expect(tree).toMatchSnapshot();
});

