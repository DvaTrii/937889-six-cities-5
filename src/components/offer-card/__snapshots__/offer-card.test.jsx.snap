// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should OfferCard render properly 1`] = `
<article
  className="cities__place-card place-card"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="place-card__mark"
  >
    <span>
      Premium
    </span>
  </div>
  <div
    className="cities__image-wrapper place-card__image-wrapper"
  >
    <a
      href="#"
    >
      <img
        alt="Place image"
        className="place-card__image"
        height="200"
        src="img/apartment-02.jpg"
        width="260"
      />
    </a>
  </div>
  <div
    className=" place-card__info"
  >
    <div
      className="place-card__price-wrapper"
    >
      <div
        className="place-card__price"
      >
        <b
          className="place-card__price-value"
        >
          €
          100
        </b>
        <span
          className="place-card__price-text"
        >
          / night
        </span>
      </div>
      <button
        className="place-card__bookmark-button button
              place-card__bookmark-button--active"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="place-card__bookmark-icon"
          height="19"
          width="18"
        >
          <use
            xlinkHref="#icon-bookmark"
          />
        </svg>
        <span
          className="visually-hidden"
        >
          In bookmarks
        </span>
      </button>
    </div>
    <div
      className="place-card__rating rating"
    >
      <div
        className="place-card__stars rating__stars"
      >
        <span
          style={
            Object {
              "width": "100%",
            }
          }
        />
        <span
          className="visually-hidden"
        >
          Rating
        </span>
      </div>
    </div>
    <h2
      className="place-card__name"
    >
      <a
        href="/offer/101"
        onClick={[Function]}
      >
        Wood and stone place
      </a>
    </h2>
    <p
      className="place-card__type"
    >
      Apartment
    </p>
  </div>
</article>
`;
